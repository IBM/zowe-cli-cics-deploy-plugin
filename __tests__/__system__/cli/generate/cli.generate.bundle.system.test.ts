/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
* Copyright IBM Corp, 2019
*
*/

import * as fse from "fs-extra";
import * as path from "path";
import {runCliScript} from "../../../__src__/TestUtils";
import {ITestEnvironment} from "../../../__src__/environment/doc/response/ITestEnvironment";
import {TestEnvironment} from "../../../__src__/environment/TestEnvironment";

// Test environment will be populated in the "beforeAll"
let TEST_ENVIRONMENT: ITestEnvironment;
let TEST_APPS_DIR: string;
let SIMPLE_TEST_APP: string;
let NO_PACKAGE_JSON_APP: string;


describe("cics-deploy generate bundle", () => {

    // Create the unique test environment
    beforeAll(async () => {
        TEST_ENVIRONMENT = await TestEnvironment.setUp({
            testName: "generate_bundle_command",
            installPlugin: true,
            tempProfileTypes: ["zosmf"]
        });
        TEST_APPS_DIR = path.join(TEST_ENVIRONMENT.workingDir, "apps");
        SIMPLE_TEST_APP = path.join(TEST_APPS_DIR, "simple-node-app");
        NO_PACKAGE_JSON_APP = path.join(TEST_APPS_DIR, "no-package-json");
    });

    afterAll(async () => {
        await TestEnvironment.cleanUp(TEST_ENVIRONMENT);
    });

    beforeEach(() => {
        if (fse.existsSync(TEST_APPS_DIR)) {
            fse.removeSync(TEST_APPS_DIR);
        }
        fse.mkdirSync(TEST_APPS_DIR);
        fse.copySync(__dirname + "../../../../__resources__/apps", TEST_APPS_DIR);
    });

    it("should generate a bundle using defaults from package.json", async () => {
        await testBundleGenerateWorks([]);
        expectFileToMatchSnapshot(path.join(SIMPLE_TEST_APP, "nodejsapps", "cics-nodejs-invoke.profile"));
    });

    it("should generate an empty bundle if no package.json present", async () => {
        const response = await runCliScript(__dirname + "/__scripts__/generate_bundle.sh", TEST_ENVIRONMENT, [NO_PACKAGE_JSON_APP]);

        expect(response.stderr.toString()).toBe("");
        expect(response.status).toBe(0);
        expectFileToMatchSnapshot(path.join(NO_PACKAGE_JSON_APP, "META-INF", "cics.xml"));
        expect(fse.existsSync(path.join(NO_PACKAGE_JSON_APP, "nodejsapps"))).toBeFalsy();
    });

    describe("paramters", async () => {
        it("should customise bundle ID according to command line args", async () => {
            await testBundleGenerateWorks(["--bundleid", "myNodeBundle"], "myNodeBundle");
        });
        it("should customise bundle version according to command line args", async () => {
            await testBundleGenerateWorks(["--bundleversion", "2.1.5"]);
        });
        it("should customise NODEJSAPP name according to command line args", async () => {
            await testBundleGenerateWorks(["--nodejsapp", "myapp"], "myapp");
        });
        it("should customise NODEJSAPP name and bundle ID independently", async () => {
            await testBundleGenerateWorks(["--bundleid", "mybundle", "--nodejsapp", "myapp"], "myapp");
        });
        it("should customise startscript according to args", async () => {
            await testBundleGenerateWorks(["--startscript", "other.js"], "cics-nodejs-invoke");
        });

        it("should customise port in generated profile accord to args", async () => {
            await testBundleGenerateWorks(["--port", "12345"]);
            expectFileToMatchSnapshot(path.join(SIMPLE_TEST_APP, "nodejsapps", "cics-nodejs-invoke.profile"));
        });

        it("should generate a bundle using values supplied when there's no package.json", async () => {
            await testBundleGenerateWorks(["--bundleid", "mybundle", "--nodejsapp", "myapp", "--startscript", "server.js"], "myapp", NO_PACKAGE_JSON_APP);
        });

        it("should return an error if invalid bundle version is supplied", async () => {
            await expectError(SIMPLE_TEST_APP, ["--bundleversion", "foo.bar.baz"]);
        });

        it("should mangle NODEJSAPP name supplied on command line", async () => {
            await testBundleGenerateWorks(["--nodejsapp", "foo%€@"], "fooXXX");
        });

        it("should mangle bundle ID supplied on command line", async () => {
            await testBundleGenerateWorks(["--bundleid", "foo%€@"], "fooXXX");
        });
    });

    describe("package.json variations", async () => {
        it("should mangle name supplied in package.json", async () => {
            await testBundleGenerateWorks([], "thisisareallylongnamethatwillnee", path.join(TEST_APPS_DIR, "long-name"));
        });

        it("should error nicely if package.json is malformed", async () => {
            const appPath = path.join(TEST_APPS_DIR, "bad-package-json");
            await expectError(appPath);
        });

        it("should error nicely if package.json is empty", async () => {
            const appPath = path.join(TEST_APPS_DIR, "empty-package-json");
            await expectError(appPath);
        });

        it("should error nicely if package.json doesn't have start script or main", async () => {
            const appPath = path.join(TEST_APPS_DIR, "minimal-package-json");
            await expectError(appPath);
        });
    });
});

async function expectError(appPath: string, options: string[] = []) {
    const response = await runCliScript(__dirname + "/__scripts__/generate_bundle.sh", TEST_ENVIRONMENT, [appPath].concat(options));
    expect(fse.existsSync(path.join(appPath, "nodejsapps"))).toBeFalsy();
    expect(response.stdout.toString()).toBe("");
    expect(response.stderr.toString()).not.toBe("");
    // expect(response.stderr.toString()).toBe("foo");
    expect(response.status).toBeGreaterThan(0);
}

async function testBundleGenerateWorks(args: string[], nodejsappName = "cics-nodejs-invoke", appPath = SIMPLE_TEST_APP) {
    const response = await runCliScript(__dirname + "/__scripts__/generate_bundle.sh", TEST_ENVIRONMENT, [appPath].concat(args));
    expect(response.stderr.toString()).toBe("");
    expect(response.status).toBe(0);
    expectFileToMatchSnapshot(path.join(appPath, "META-INF", "cics.xml"));
    expectFileToMatchSnapshot(path.join(appPath, "nodejsapps", nodejsappName + ".nodejsapp"));
}

function expectFileToMatchSnapshot(filePath: string) {
    const contents = fse.readFileSync(filePath).toString();
    expect(contents).toMatchSnapshot();
}

