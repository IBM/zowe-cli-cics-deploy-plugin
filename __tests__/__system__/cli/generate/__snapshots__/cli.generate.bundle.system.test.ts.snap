// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cics-deploy generate bundle package.json variations should mangle name supplied in package.json 1`] = `
"<manifest xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle\\" bundleVersion=\\"1\\" bundleRelease=\\"0\\" id=\\"thisisareallylongnamethatwillneedtobetruncatedinordertofitinabun\\" bundleMajorVer=\\"1\\" bundleMinorVer=\\"0\\" bundleMicroVer=\\"0\\"><define name=\\"thisisareallylongnamethatwillnee\\" type=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" path=\\"nodejsapps/thisisareallylongnamethatwillnee.nodejsapp\\"></define></manifest>
"
`;

exports[`cics-deploy generate bundle package.json variations should mangle name supplied in package.json 2`] = `
"<nodejsapp xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" name=\\"thisisareallylongnamethatwillnee\\" startscript=\\"server.js\\" profile=\\"nodejsapps/thisisareallylongnamethatwillnee.profile\\" lerunopts=\\"DFHSJNRO\\"></nodejsapp>
"
`;

exports[`cics-deploy generate bundle paramters should customise NODEJSAPP name according to command line args 1`] = `
"<manifest xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle\\" bundleVersion=\\"1\\" bundleRelease=\\"0\\" id=\\"cics-nodejs-invoke\\" bundleMajorVer=\\"1\\" bundleMinorVer=\\"0\\" bundleMicroVer=\\"0\\"><define name=\\"myapp\\" type=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" path=\\"nodejsapps/myapp.nodejsapp\\"></define></manifest>
"
`;

exports[`cics-deploy generate bundle paramters should customise NODEJSAPP name according to command line args 2`] = `
"<nodejsapp xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" name=\\"myapp\\" startscript=\\"server.js\\" profile=\\"nodejsapps/myapp.profile\\" lerunopts=\\"DFHSJNRO\\"></nodejsapp>
"
`;

exports[`cics-deploy generate bundle paramters should customise NODEJSAPP name and bundle ID independently 1`] = `
"<manifest xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle\\" bundleVersion=\\"1\\" bundleRelease=\\"0\\" id=\\"mybundle\\" bundleMajorVer=\\"1\\" bundleMinorVer=\\"0\\" bundleMicroVer=\\"0\\"><define name=\\"myapp\\" type=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" path=\\"nodejsapps/myapp.nodejsapp\\"></define></manifest>
"
`;

exports[`cics-deploy generate bundle paramters should customise NODEJSAPP name and bundle ID independently 2`] = `
"<nodejsapp xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" name=\\"myapp\\" startscript=\\"server.js\\" profile=\\"nodejsapps/myapp.profile\\" lerunopts=\\"DFHSJNRO\\"></nodejsapp>
"
`;

exports[`cics-deploy generate bundle paramters should customise bundle ID according to command line args 1`] = `
"<manifest xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle\\" bundleVersion=\\"1\\" bundleRelease=\\"0\\" id=\\"myNodeBundle\\" bundleMajorVer=\\"1\\" bundleMinorVer=\\"0\\" bundleMicroVer=\\"0\\"><define name=\\"myNodeBundle\\" type=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" path=\\"nodejsapps/myNodeBundle.nodejsapp\\"></define></manifest>
"
`;

exports[`cics-deploy generate bundle paramters should customise bundle ID according to command line args 2`] = `
"<nodejsapp xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" name=\\"myNodeBundle\\" startscript=\\"server.js\\" profile=\\"nodejsapps/myNodeBundle.profile\\" lerunopts=\\"DFHSJNRO\\"></nodejsapp>
"
`;

exports[`cics-deploy generate bundle paramters should customise bundle version according to command line args 1`] = `
"<manifest xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle\\" bundleVersion=\\"1\\" bundleRelease=\\"0\\" bundleMajorVer=\\"2\\" bundleMinorVer=\\"1\\" bundleMicroVer=\\"5\\" id=\\"cics-nodejs-invoke\\"><define name=\\"cics-nodejs-invoke\\" type=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" path=\\"nodejsapps/cics-nodejs-invoke.nodejsapp\\"></define></manifest>
"
`;

exports[`cics-deploy generate bundle paramters should customise bundle version according to command line args 2`] = `
"<nodejsapp xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" name=\\"cics-nodejs-invoke\\" startscript=\\"server.js\\" profile=\\"nodejsapps/cics-nodejs-invoke.profile\\" lerunopts=\\"DFHSJNRO\\"></nodejsapp>
"
`;

exports[`cics-deploy generate bundle paramters should customise port in generated profile accord to args 1`] = `
"<manifest xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle\\" bundleVersion=\\"1\\" bundleRelease=\\"0\\" id=\\"cics-nodejs-invoke\\" bundleMajorVer=\\"1\\" bundleMinorVer=\\"0\\" bundleMicroVer=\\"0\\"><define name=\\"cics-nodejs-invoke\\" type=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" path=\\"nodejsapps/cics-nodejs-invoke.nodejsapp\\"></define></manifest>
"
`;

exports[`cics-deploy generate bundle paramters should customise port in generated profile accord to args 2`] = `
"<nodejsapp xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" name=\\"cics-nodejs-invoke\\" startscript=\\"server.js\\" profile=\\"nodejsapps/cics-nodejs-invoke.profile\\" lerunopts=\\"DFHSJNRO\\"></nodejsapp>
"
`;

exports[`cics-deploy generate bundle paramters should customise port in generated profile accord to args 3`] = `
"# This is a profile for a CICS NODEJSAPP resource

# Import the provisioned configuration file for this NODEJSAPP,
# this file must be created before this Bundle can be installed in CICS
INCLUDE=&CONFIGROOT;/nodejsapps/cics-nodejs-invoke.included.profile

# Set the PORT envionment variable, application code should reference this
# value in preference to a hard-coded port number, the value references an
# environment variable that will be configured within the provisioned configuration file.
PORT=12345
"
`;

exports[`cics-deploy generate bundle paramters should customise startscript according to args 1`] = `
"<manifest xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle\\" bundleVersion=\\"1\\" bundleRelease=\\"0\\" id=\\"cics-nodejs-invoke\\" bundleMajorVer=\\"1\\" bundleMinorVer=\\"0\\" bundleMicroVer=\\"0\\"><define name=\\"cics-nodejs-invoke\\" type=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" path=\\"nodejsapps/cics-nodejs-invoke.nodejsapp\\"></define></manifest>
"
`;

exports[`cics-deploy generate bundle paramters should customise startscript according to args 2`] = `
"<nodejsapp xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" name=\\"cics-nodejs-invoke\\" startscript=\\"other.js\\" profile=\\"nodejsapps/cics-nodejs-invoke.profile\\" lerunopts=\\"DFHSJNRO\\"></nodejsapp>
"
`;

exports[`cics-deploy generate bundle paramters should generate a bundle using values supplied when there's no package.json 1`] = `
"<manifest xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle\\" bundleVersion=\\"1\\" bundleRelease=\\"0\\" id=\\"mybundle\\"><define name=\\"myapp\\" type=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" path=\\"nodejsapps/myapp.nodejsapp\\"></define></manifest>
"
`;

exports[`cics-deploy generate bundle paramters should generate a bundle using values supplied when there's no package.json 2`] = `
"<nodejsapp xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" name=\\"myapp\\" startscript=\\"server.js\\" profile=\\"nodejsapps/myapp.profile\\" lerunopts=\\"DFHSJNRO\\"></nodejsapp>
"
`;

exports[`cics-deploy generate bundle paramters should mangle NODEJSAPP name supplied on command line 1`] = `
"<manifest xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle\\" bundleVersion=\\"1\\" bundleRelease=\\"0\\" id=\\"cics-nodejs-invoke\\" bundleMajorVer=\\"1\\" bundleMinorVer=\\"0\\" bundleMicroVer=\\"0\\"><define name=\\"fooXXX\\" type=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" path=\\"nodejsapps/fooXXX.nodejsapp\\"></define></manifest>
"
`;

exports[`cics-deploy generate bundle paramters should mangle NODEJSAPP name supplied on command line 2`] = `
"<nodejsapp xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" name=\\"fooXXX\\" startscript=\\"server.js\\" profile=\\"nodejsapps/fooXXX.profile\\" lerunopts=\\"DFHSJNRO\\"></nodejsapp>
"
`;

exports[`cics-deploy generate bundle paramters should mangle bundle ID supplied on command line 1`] = `
"<manifest xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle\\" bundleVersion=\\"1\\" bundleRelease=\\"0\\" id=\\"fooXXX\\" bundleMajorVer=\\"1\\" bundleMinorVer=\\"0\\" bundleMicroVer=\\"0\\"><define name=\\"fooXXX\\" type=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" path=\\"nodejsapps/fooXXX.nodejsapp\\"></define></manifest>
"
`;

exports[`cics-deploy generate bundle paramters should mangle bundle ID supplied on command line 2`] = `
"<nodejsapp xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" name=\\"fooXXX\\" startscript=\\"server.js\\" profile=\\"nodejsapps/fooXXX.profile\\" lerunopts=\\"DFHSJNRO\\"></nodejsapp>
"
`;

exports[`cics-deploy generate bundle should generate a bundle using defaults from package.json 1`] = `
"<manifest xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle\\" bundleVersion=\\"1\\" bundleRelease=\\"0\\" id=\\"cics-nodejs-invoke\\" bundleMajorVer=\\"1\\" bundleMinorVer=\\"0\\" bundleMicroVer=\\"0\\"><define name=\\"cics-nodejs-invoke\\" type=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" path=\\"nodejsapps/cics-nodejs-invoke.nodejsapp\\"></define></manifest>
"
`;

exports[`cics-deploy generate bundle should generate a bundle using defaults from package.json 2`] = `
"<nodejsapp xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP\\" name=\\"cics-nodejs-invoke\\" startscript=\\"server.js\\" profile=\\"nodejsapps/cics-nodejs-invoke.profile\\" lerunopts=\\"DFHSJNRO\\"></nodejsapp>
"
`;

exports[`cics-deploy generate bundle should generate a bundle using defaults from package.json 3`] = `
"# This is a profile for a CICS NODEJSAPP resource

# Import the provisioned configuration file for this NODEJSAPP,
# this file must be created before this Bundle can be installed in CICS
INCLUDE=&CONFIGROOT;/nodejsapps/cics-nodejs-invoke.included.profile

# Set the PORT envionment variable, application code should reference this
# value in preference to a hard-coded port number, the value references an
# environment variable that will be configured within the provisioned configuration file.
PORT=&HTTP_PORT;
"
`;

exports[`cics-deploy generate bundle should generate an empty bundle if no package.json present 1`] = `
"<manifest xmlns=\\"http://www.ibm.com/xmlns/prod/cics/bundle\\" bundleVersion=\\"1\\" bundleRelease=\\"0\\"></manifest>
"
`;
